src/logger/src/logger.rs

```diff
@@ -541,130 +550,189 @@ mod tests {
-        l.set_include_origin(false, true);
-        assert_eq!(l.show_line_numbers(), false);
+    fn test_configure() {
+        let logger = Logger::new();
+
+        // Assert that `configure()` can be called successfully any number of times.
+        assert!(logger.configure(Some(TEST_INSTANCE_ID.to_string())).is_ok());
+        assert!(logger.configure(None).is_ok());
+        assert!(logger.configure(Some(TEST_INSTANCE_ID.to_string())).is_ok());
+
+        // Assert that `init()` works after `config()`
+        let (writer, mut reader) = log_channel();
+        logger.set_max_level(log::LevelFilter::Info);
+        assert!(logger
+            .init(TEST_APP_HEADER.to_string(), Box::new(writer))
+            .is_ok());
+        validate_log(Box::new(&mut reader), &format!("{}\n", TEST_APP_HEADER));
+        // Check that the logs are written to the configured writer.
+        log(&logger, Level::Info, "info");
+        validate_log(
+            Box::new(&mut reader),
+            "[TEST-INSTANCE-ID:INFO:logger.rs:0] info\n",
+        );
+    }
 
-        l.set_include_origin(true, true)
-            .set_include_level(true)
-            .set_max_level(log::LevelFilter::Info);
-        assert_eq!(l.show_line_numbers(), true);
-        assert_eq!(l.show_file_path(), true);
-        assert_eq!(l.show_level(), true);
+    #[test]
+    fn test_init() {
+        let logger = Logger::new();
+
+        // Assert that the first call to `init()` is successful.
+        let (writer, mut reader) = log_channel();
+        logger.set_max_level(log::LevelFilter::Info);
+        logger.set_instance_id(TEST_INSTANCE_ID.to_string());
+        assert!(logger
+            .init(TEST_APP_HEADER.to_string(), Box::new(writer))
+            .is_ok());
+        validate_log(Box::new(&mut reader), &format!("{}\n", TEST_APP_HEADER));
+        // Check that the logs are written to the configured writer.
+        log(&logger, Level::Info, "info");
+        validate_log(
+            Box::new(&mut reader),
+            "[TEST-INSTANCE-ID:INFO:logger.rs:0] info\n",
+        );
 
-        l.set_instance_id(TEST_INSTANCE_ID.to_string());
+        // Assert that initialization works only once.
+        let (writer_2, mut reader_2) = log_channel();
+        assert!(logger
+            .init(TEST_APP_HEADER.to_string(), Box::new(writer_2),)
+            .is_err());
+        // Check that the logs are written only to the first writer.
+        log(&logger, Level::Info, "info");
+        validate_log(
+            Box::new(&mut reader),
+            "[TEST-INSTANCE-ID:INFO:logger.rs:0] info\n",
+        );
+        validate_log(Box::new(&mut reader_2), "");
+    }
 
-        // Assert that the initial configuration works any number of times.
-        assert!(l.configure(Some(TEST_INSTANCE_ID.to_string())).is_ok());
-        assert!(l.configure(None).is_ok());
-        assert!(l.configure(Some(TEST_INSTANCE_ID.to_string())).is_ok());
+    #[test]
+    fn test_create_prefix() {
```

Comment:
I noticed that now you are changing also the level and yep, 4 parameters would be indeed a lot.
But why do you modify the level for testing `create_prefix()`? They don't seem to be related. Anyway I would still try to reduce this test, either by creating a function that takes the 3 flags as parameters or by removing some of the tested scenarios: for example, I don't see why we should test three times the mock instance id (the `:` join separator is already super tested before).

Reply:
The level is logged as part of the prefix. For example: 

```
"[TEST-INSTANCE-ID:INFO:logger.rs:0] log"
```

or

```
"[TEST-INSTANCE-ID:ERROR:logger.rs:0] log"
```

So we should test that it's logged correctly.

The function would have to take more then 3 parameters: 
1. logger
2. reader
3. log level
4. show_level
5. show_file
6. show_line_number
7. expected string

Taking this into consideration I don't think a function would be very readable. 

I removed some test cases.

Reply:
The level is logged as part of the prefix indeed but the `create_prefix()` behavior is not different based on the level value or is it? (the function just checks if `show_level` is set, does `show_level()` behave differently based on the level value?).
For the separate function, I didn't want to include also the `validate_log()` call.
Anyway, looks good now!

Reply:
The level doesn't change the behavior of `create_prefix()`, but there might be a bug inside `create_prefix()` where it would directly print `INFO` instead of the actual level. I think it brings value checking the level.